test-runner-poc/lib/runner.js

```diff
@@ -2,72 +2,135 @@
 
 const path = require("path");
 const fs = require("fs");
-const utils = require("mocha").utils;
+const MongoClient = require('mongodb').MongoClient;
+const environments = require('../environments');
+const TestConfiguration = require('../config');
+const parseConnectionString = require('../../lib/core/uri_parser');
+const mock = require('mongodb-mock-server');
 
+let mongoClient;
 let filters = [];
 let files = [];
 
 function addFilter(filter) {
-	if (typeof filter !== "function" && typeof filter !== "object") {
-		throw new Error(
-			"Type of filter must either be a function or an object"
-		);
-	}
-	if (
-		typeof filter === "object" &&
-		(!filter.filter || typeof filter.filter !== "function")
-	) {
-		throw new Error("Object filters must have a function named filter");
-	}
-
-	if (typeof filter === "function") {
-		filters.push({ filter: filter });
-	} else {
-		filters.push(filter);
-	}
+  switch (typeof filter) {
+    case 'function':
+      filters.push({ filter: filter });
+      break;
+    case 'object':
+      if (!filter.filter || typeof filter.filter !== 'function') {
+        throw new Error('Object filters must have a function named filter');
+      }
+      filters.push(filter);
+      break;
+    default:
+      throw new Error('Type of filter must either be a function or an object');
+  }
 }
 
-before(function() {
+function environmentSetup(environmentCallback) {
+  const mongodb_uri = process.env.MONGODB_URI || 'mongodb://localhost:27017';
+  mongoClient = new MongoClient(mongodb_uri);
 
-	const environmentName = process.env['MONGODB_ENVIRONMENT'];
+  mongoClient.connect((err, client) => {
+    if (err) throw new Error(err);
 
-	console.log(`[environment: ${environmentName}]`);
+    createFilters(environmentParser);
 
-	//apply filters
-	fs.readdirSync(path.join(__dirname, "filters"))
-		.filter(x => x.indexOf("js") !== -1)
-		.forEach(x => {
-			const FilterModule = require(path.join(__dirname, "filters", x));
-			addFilter(new FilterModule({ runtimeTopology: 'sharded' }));
-		});
+    function environmentParser(environmentName, version) {
+      console.log('environmentName ', environmentName);
+      const Environment = environments[environmentName];
 
+      let host, port;
+      parseConnectionString(mongodb_uri, (err, parsedURI) => {
+        if (err) throw new Error(err);
+        port = parsedURI.hosts[0].port;
+        host = parsedURI.hosts[0].host;
+
```

Comment:
`mongodb_uri` parsing now happens before constructing the `environment` and then the `mongoPackage` is resolved.